src/lib/raft/raft.c

```diff
@@ -206,10 +206,51 @@ raft_scores_str(const struct raft *raft)
 	return buf;
 }
 
+static inline bool
+raft_is_leader_seen(const struct raft *raft)
+{
+	bool is_seen = bit_test(&raft->leader_witness_map, raft->self);
+	assert(!is_seen || raft->leader != 0);
+	return is_seen;
+}
+
 /** Schedule broadcast of the complete Raft state to all the followers. */
 static void
 raft_schedule_broadcast(struct raft *raft);
 
+/**
+ * Update all leader-related fields on leader appearance and let others know the
+ * leader is seen.
+ */
+static inline void
+raft_set_leader(struct raft *raft, uint32_t leader)
+{
+	assert(raft->leader == 0);
+	assert(leader != 0);
+	raft->leader = leader;
+	bool is_changed = raft_notify_is_leader_seen(raft, true, raft->self);
+	assert(is_changed);
+	(void)is_changed;
+	raft->leader_last_seen = raft_ev_now(raft_loop());
+	/* Leader is seen. Tell other nodes about its appearance. */
+	raft_schedule_broadcast(raft);
+}
+
+/**
+ * Clear leader-related fields on leader loss or resign and let everyone know
+ * this node doesn't see the leader anymore.
+ */
+static inline void
+raft_clear_leader(struct raft *raft)
```

Comment:
Yes, because we want to update 'leader_last_seen' when a node is a leader and then it receives a term bump from not a leader. Everything else (clearing witness map and a broadcast) is already done in`raft_sm_schedule_new_term`

Thanks for paying attention!